### PR TITLE
fix(tianmu): Fix delete from Multiple-Table, Failed to delete the data (#663)

### LINKED ISSUE #663: bug: delete from Multiple-Table, Failed to delete the data
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

Failed to delete the data!
image

### Expected behavior

image

### How To Reproduce

```sql
create table t11 (a int NOT NULL, b int, primary key (a))ENGINE=TIANMU;
create table t12 (a int NOT NULL, b int, primary key (a))ENGINE=TIANMU;
insert into t11 values (0, 10),(1, 11),(2, 12);
insert into t12 values (33, 10),(0, 11),(2, 12);

select * from t11,t12;
delete t11.*,t12.* from t11,t12 where t11.a = t12.a;
```

### Environment

ubuntu@ubuntu:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

ubuntu@ubuntu:~$ /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld Ver 5.7.36_v1.0.1_beta_2-StoneDB for Linux on x86_64 (build-)
build information as follow:
Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
Branch name: stonedb-5.7-dev
Last commit ID: https://github.com/stoneatom/stonedb/commit/208f79d80bae1667d147e0713950317c9ca6c6b9
Last commit time: Date: Sat Sep 17 16:19:42 2022 +0800
Build time: Date: Fri Sep 23 14:14:27 CST 2022

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by konghaiya**
fix(tianmu)Fix delete from Multiple-Table, Failed to delete the data (#663)(#678)
Cause:
At present, the primary key of the tianmu engine does not fully support delete and update statements. 
Temporary solution:
Modify the execution plan of the SQL layer so that the delete and update statements do not follow the primary key logic

fix issue: 
- #663 
- #678

## Summary about this PR

Issue Number: close #663 , #678

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Comment by isredstar** [comment]
Will the solution affect other queries？ @konghaiya

**Comment by konghaiya** [comment]
The query will not be affected. The modification here only involves the delete and update statements of tianmu.

File: mysql-test/suite/tianmu/r/delete.result [hunk 1]
```diff
@@ -198,19 +198,16 @@ select * from t11;
 a	b
 0	10
 1	11
-2	12
 select * from t12;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/delete.result:201]: The PR title format

File: mysql-test/suite/tianmu/t/issue663.test [hunk 2]
```diff
@@ -0,0 +1,32 @@
+--source include/have_tianmu.inc
+
+--echo #
+--echo # Fix delete from Multiple-Table, Failed to delete the data #663,#678
+--echo #
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue663.test]: delete unrelated comment messages

File: mysql-test/suite/tianmu/t/issue663.test [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+drop table t11,t12;
+
+CREATE TABLE t1 (a int not null,b int not null)ENGINE=TIANMU;
+CREATE TABLE t2 (a int not null, b int not null, primary key (a,b))ENGINE=TIANMU;
+CREATE TABLE t3 (a int not null, b int not null, primary key (a,b))ENGINE=TIANMU;
+insert into t1 values (1,1),(2,1),(1,3);
+insert into t2 values (1,1),(2,2),(3,3);
+insert into t3 values (1,1),(2,1),(1,3);
+delete t2.*,t3.* from t1,t2,t3 where t1.a=t2.a AND t2.b=t3.a and t1.b=t3.b;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue663.test:28]: Addings more `where` conditions, such as combinations conditions.

File: mysql-test/suite/tianmu/t/issue663.test [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+select * from t1;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue663.test:30]: `select` statement should be test  with differ selection conditions.

File: mysql-test/suite/tianmu/t/issue663.test [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+select * from t2;
+select * from t3;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue663.test:32]: And tests for query plan should be existed.

File: sql/sql_planner.cc [hunk 6]
```diff
@@ -954,7 +954,18 @@ void Optimize_table_order::best_access_path(JOIN_TAB *tab,
...
   uint used_key_parts= 0;
 
-  if (tab->keyuse() != NULL)
+  /*
+    The primary key of the tianmu engine does not support delete and update statements.
+    The following codes can be deleted after subsequent support
+  */
+  bool tianmu_engine = table->s->db_type() ? table->s->db_type()->db_type == DB_TYPE_TIANMU: false;
+  enum_sql_command sqlCommand = table->in_use->lex->sql_command;
+  bool tianmuDeleteOrUpdate = (tianmu_engine && (sqlCommand == SQLCOM_DELETE ||
```
Comment (RingsC) [inline sql/sql_planner.cc]: I think this logic is better in optimization. in optimization, if we foud the table is Tianmu table. we disable the `key` information. 

Then, here, tab->keyuse() of a Tianmu table is in disable status.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc [hunk 7]
```diff
@@ -5126,7 +5126,19 @@ bool JOIN::make_join_plan()
   // Build the key access information, which is the basis for ref access.
   if (where_cond || select_lex->outer_join)
   {
-    if (update_ref_and_keys(thd, &keyuse_array, join_tab, tables, where_cond,
+    /*
+      The primary key of the tianmu engine does not support delete and update statements.
+      The following codes can be deleted after subsequent support
+    */
+    TABLE *const table= join_tab->table();
```
Comment (RingsC) [inline sql/sql_optimizer.cc:5133]: Adds a checkt to test it is nullptr or not.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc (hunk 7 above)
Comment (isredstar) [reply sql/sql_optimizer.cc:5133]: It is necessary @konghaiya

Comment (konghaiya) [reply sql/sql_optimizer.cc:5133]: Code for nullptr check has been added

**Review by RingsC — APPROVED** [review]
LGTM

**Review by isredstar — APPROVED** [review]

**Comment by RingsC** [comment]
Now, tianmu, is not support PK very well, we will improve to support PK in future. And, that time, we will remove this optimization restriction.

**Pull request merged**